src/Development/IDE/Core/Shake.hs

```diff
@@ -375,57 +379,114 @@ shakeProfile :: IdeState -> FilePath -> IO ()
         )
         -- It is crucial to be masked here, otherwise we can get killed
-        -- between spawning the new thread and updating shakeAbort.
+        -- between spawning the new thread and updating shakeSession.
         -- See https://github.com/digital-asset/ghcide/issues/79
-        (do
-              start <- offsetTime
-              aThread <- asyncWithUnmask $ \restore -> do
-                   res <- try (restore $ shakeRunDatabaseProfile shakeProfileDir shakeDb acts)
-                   runTime <- start
-                   let res' = case res of
-                            Left e -> "exception: " <> displayException e
-                            Right _ -> "completed"
-                       profile = case res of
-                            Right (_, Just fp) ->
-                                let link = case filePathToUri' $ toNormalizedFilePath' fp of
-                                                NormalizedUri _ x -> x
-                                in ", profile saved at " <> T.unpack link
-                            _ -> ""
-                   let logMsg = logDebug logger $ T.pack $
-                        "Finishing shakeRun (took " ++ showDuration runTime ++ ", " ++ res' ++ profile ++ ")"
-                   return (fst <$> res, logMsg)
-              let wrapUp (res, _) = do
-                    either (throwIO @SomeException) return res
-              _ <- async $ do
-                  (_, logMsg) <- wait aThread
-                  logMsg
-              pure (cancel aThread, wrapUp =<< wait aThread))
+        (\() -> newSession it acts)
+
+-- | Enqueue an action in the existing 'ShakeSession'.
+--   Returns a computation to block until the action is run.
+--   Assumes a 'ShakeSession' is available.
+--
+--   Appropriate for user actions other than edits.
+shakeEnqueue :: IdeState -> Action a -> IO (IO a)
+shakeEnqueue IdeState{shakeSession} act =
+    withMVar shakeSession $ \s -> runInShakeSession s act
+
+-- Set up a new 'ShakeSession' with a set of initial actions.
+-- Will crash if there is an existing 'ShakeSession' running.
+-- Reports progress on the set of initial actions only.
+newSession :: IdeState -> [Action a] -> IO (ShakeSession, IO [a])
+newSession IdeState{shakeExtras=ShakeExtras{..}, ..} acts = do
+    -- A work queue for actions added via 'runInShakeSession'
+    actionQueue :: TQueue (Action ()) <- newTQueueIO
+
+    start <- offsetTime
+    let
+        -- A daemon-like action used to inject additional work
+        -- Runs actions from the work queue sequentially
+        pumpAction =
+            forever $ join $ liftIO $ atomically $ readTQueue actionQueue
+
+        progressRun
+          | reportProgress = lspShakeProgress getLspId eventer inProgress
+          | otherwise = return ()
+
+        workRun restore = withAsync progressRun $ \progressThread -> do
+          let acts' =
+                [ [] <$ pumpAction
+                , parallel acts
```

Comment:
This call to `parallel` worries me. If it has negative performance implications, `lspShakeProgress` could be changed to auto-detect completion when the number of items `inProgress` is 0.

Reply:
I haven't noticed any noticeable performance degradation due to `parallel`

Reply:
It doesn't massively worry me.